### ISSUE TITLE
multi gpu train sdxl

### ISSUE THREAD

**Issue by bycodename** [kohya-ss/sd-scripts#1118]
I think there is a need for a neat multi GPU documentation now. We cannot access any information about how it should be used, what parameters, how step and epoch should be calculated, and what logic it works on. We are trying to understand through issues and conversations, but we cannot get a healthy result. Please explain this topic a little more specifically and give more support to multi GPU, which is now a great need.

**Comment by dill-shower** [comment]
> We cannot access any information about how it should be used, what parameters, how step and epoch should be calculated, and what logic it works on.

You can check this in documentations for EveryDream.

**Comment by bycodename** [comment]
Thank you for the suggestion, I just checked but couldn't find any relevant
information. I hope I explained correctly that my question is for Sd xl
multi gpu, as I stated in the title.

**Comment by dill-shower** [comment]
> any relevant information

There is no particular difference between sdxl and sd2.1 training in terms of hyperparameters. Regarding the change in lr and the rest depending on the number of GPUs, it is logical to look at the section where it is written how to calculate lr depending on the batch size. Adding a GPU simply allows you to increase the batch size and is unlikely to require any special changes to the hyperparameters

**Comment by BootsofLagrangian** [comment]
sd-scripts uses accelerate which supports DDP, DeepSpeed, FSDP etc.. When you use multi-gpu environment, batch will be multiplied by number of your gpus. For example, if you use 4 gpus and 4 train batch size, then accelerate uses total 16 batch size = 4 gpus x 4 train batch size. In sd-scripts, train batch size means "train batch size per one device." 

You maybe consider larger learning rate when you use multi-gpus because it will impact on batch size.

Other training details may not influenced on. sd1.5, sd2.1 sdxl, whatever.

**Comment by BootsofLagrangian** [comment]
> > sd-scripts uses accelerate which supports DDP, DeepSpeed, FSDP etc.. When you use multi-gpu environment, batch will be multiplied by number of your gpus. For example, if you use 4 gpus and 4 train batch size, then accelerate uses total 16 batch size = 4 gpus x 4 train batch size. In sd-scripts, train batch size means "train batch size per one device."
> > You maybe consider larger learning rate when you use multi-gpus because it will impact on batch size.
> > Other training details may not influenced on. sd1.5, sd2.1 sdxl, whatever.
> 
> Then, as far as I understand, I need to adjust the parameters by dividing the epochs and steps by the number of GPUs, is it true or are there other parameters I missed? But even if I set it as 1 epoch, it is okay to say that there will be as many epochs as the number of GPUs, but why then, for example, when there are 4 GPUs, I divide the 4k steps by 4 and set it as 1k steps. Then shouldn't it do the 1k step on all 4 GPUs at the same time and finish the process? Why does it start the second 1k step after finishing the 1k step? I'm sorry, it's a bit confusing :) I may have misunderstood ddp. But I can say that there is a serious performance decrease when compared to 1 GPU, which makes me wonder if it is an efficient enough solution. Do I need to split num_cpu_threads_per_process here as well? The parameters I use for ddp are --ddp_timeout=10000000 --ddp_gradient_as_bucket_view --ddp_static_graph, sorry again as I'm very new to this.

`Accelerate` and `sd-scripts` automatically divide `total train steps` by `number of gpus`.
If you want to set `total train steps` via `epochs`, just write how many epochs to train. Then, sd-scripts calculates `total train steps` using `train epochs` and divides them by `number of gpus`
And default setting of multi gpu training in accelerate is DDP. DDP (Distributed Data Parallel) is a method where GPUs copy parameter gradients from the optimizer state and receive divided datasets for training. So, DDP consumes slightly more time for synchronizing each states, but it is really small amount of time. I think there is something wrong in your environment
